Add `proc_macro` for generating `NciArray`s

augenfrosch opened the pull request:
Adds a procedural macro for constructing `NciArray`s at compile time. Previously, the `NciArrayBuilder` was (typically) used to generate them, but that required an explicit build script, while the macro can be used to define them directly (but is subject to other limitations).

**Example:**
```rust
pub const EXAMPLE_ARRAY: NciArray<i32, i32> = nci_array! {
    -500i32 => -500,
    -499i32 => -499,
    -2i32 => -2,
    -1i32 => -1,
    0i32 => 0,
    1i32 => 1,
    2i32 => 2,
    499i32 => 499,
    500i32 => 500,
};
```

**Caveats:**
* The index type is limited to the basic integer types. Procedural macros don't have access to (inferred) type information (and there also seems to be no way to use generics), which is necessary for the indices since the segment data has to be constructed in the macro itself. The available options are hardcoded and parsed from the token stream, which comes with the following quirks:
  * ~The integer type specific suffix has to be used to distinguish them, since otherwise what type they represent would not be known, and if the macro were to assume a default type for literals without a suffix, the final segment might extend beyond the actual index type's value range.~ No longer the case after 3ee4164ce5bcaaf117ef7f2a36418581272ff01a
  * ~It is syntactically possible to specify different integer types via the integer literal suffixes for a single array. This is not desired and is currently checked for and returns a compiler error if encountered~ Checked by Rust's type system after 3ee4164ce5bcaaf117ef7f2a36418581272ff01a
* The macro currently enforces that the entries are specified in order from the lowest to the highest index. This allows for the generation logic to be simpler and better performance, since otherwise the entries would need to be sorted

augenfrosch committed:
Add `NciIndex` implementations for `usize` and `isize`

augenfrosch committed:
Add `proc_macro` for generating `NciArray`s

augenfrosch committed:
Rename macro to `nci_array` & use curly brackets in tests

augenfrosch committed:
Add `macros` feature & move `NciIndex` to new shared crate to not create a cyclic dependency

augenfrosch committed:
Fix clippy warnings & remove left-over `dbg!`

augenfrosch committed:
Fix formatting

KonaeAkira commented on non_contiguously_indexed_array_macros/src/lib.rs:
```diff
@@ -0,0 +1,262 @@
+        match value {
+            Expr::Lit(expr_lit) => match &expr_lit.lit {
+                syn::Lit::Int(lit_int) => match lit_int.suffix() {
+                    "i8" => Ok(Self::I8(lit_int.base10_parse::<i8>()?)),
+                    "i16" => Ok(Self::I16(lit_int.base10_parse::<i16>()?)),
+                    "i32" => Ok(Self::I32(lit_int.base10_parse::<i32>()?)),
+                    "i64" => Ok(Self::I64(lit_int.base10_parse::<i64>()?)),
+                    "i128" => Ok(Self::I128(lit_int.base10_parse::<i128>()?)),
+                    "isize" => Ok(Self::Isize(lit_int.base10_parse::<isize>()?)),
+                    "u8" => Ok(Self::U8(lit_int.base10_parse::<u8>()?)),
+                    "u16" => Ok(Self::U16(lit_int.base10_parse::<u16>()?)),
+                    "u32" => Ok(Self::U32(lit_int.base10_parse::<u32>()?)),
+                    "u64" => Ok(Self::U64(lit_int.base10_parse::<u64>()?)),
+                    "u128" => Ok(Self::U128(lit_int.base10_parse::<u128>()?)),
+                    "usize" => Ok(Self::Usize(lit_int.base10_parse::<usize>()?)),
```
What about binary and hexadecimal literals?

augenfrosch replied:
It is possible to specify them as `0b1010i32` or `0xAu32` etc.
The example array can be equivalently constructed like this (after the commit I just pushed that fixes overflowing literals):
```rust
#[allow(overflowing_literals)]
const EXAMPLE_ARRAY: NciArray<i32, i32> = nci_array! {
    0xfffffe0ci32 => -500,
    0b11111111111111111111111000001101i32 => -499,
    -2i32 => -2,
    -0x1i32 => -1,
    -0b0i32 => 0,
    0x1i32 => 1,
    2i32 => 2,
    0b111110011i32 => 499,
    0x1f4i32 => 500,
};
```

augenfrosch committed:
Fix parsing of overflowing negative signed integer literals

augenfrosch committed:
Add support for byte literals

KonaeAkira commented:
You can make a new integer type to store the literals:
```rust
enum Number {
    NonNegative(u128),
    Negative(u128)
}
```

augenfrosch committed:
Allow integer literals without type suffix

augenfrosch committed:
Explicitly deny `overflowing_literals`

This avoids silently introducing inconsistencies between the parsing of literals without type suffix in the macro and by `rustc` using the inferred type.

augenfrosch committed:
Fix clippy warnings

augenfrosch committed:
Use single function call with slice argument to check types

This avoids crashing the compiler for large arrays.

augenfrosch committed:
Use `into()` instead of casting with `as`

augenfrosch committed:
Use `NciIndex::next` to detect new segments

augenfrosch committed:
Simplify index parsing error logic & output

augenfrosch commented on non_contiguously_indexed_array_macros/src/lib.rs:
```diff
@@ -0,0 +1,174 @@
+                    expr_unary.span(),
+                    "Unsupported unary operator! Only negation, `-`, is allowed.",
+                )),
+            },
```
I tried to shorten the match arm like your example (with the necessary structs) but stopped at this:
```rust
Expr::Unary(syn::ExprUnary { op: syn::UnOp::Neg(_), expr, ..}) => { /* ... */}
```
where `expr: &Box<Expr>`. I didn't find a way of pattern matching the value in `Box` that doesn't require an unstable feature.

I ended up rewriting it differently to hopefully similar effect. See 7838902db75c7759bcd3b6c3f4b9451ff0aec88a.
